libeos-parental-controls/app-filter.c

```diff
@@ -452,3 +483,510 @@ epc_get_app_filter_finish (GAsyncResult  *result,
+
+/**
+ * epc_app_filter_builder_init:
+ * @builder: an uninitialised #EpcAppFilterBuilder
+ *
+ * Initialise the given @builder so it can be used to construct a new
+ * #EpcAppFilter. @builder must have been allocated on the stack, and must not
+ * already be initialised.
+ *
+ * Construct the #EpcAppFilter by calling methods on @builder, followed by
+ * epc_app_filter_builder_end(). To abort construction, use
+ * epc_app_filter_builder_clear().
+ *
+ * Since: 0.1.0
+ */
+void
+epc_app_filter_builder_init (EpcAppFilterBuilder *builder)
+{
+  EpcAppFilterBuilder local_builder = EPC_APP_FILTER_BUILDER_INIT ();
+  EpcAppFilterBuilderReal *_builder = (EpcAppFilterBuilderReal *) builder;
+
+  g_return_if_fail (_builder != NULL);
+  g_return_if_fail (_builder->paths_blacklist == NULL);
+
+  memcpy (builder, &local_builder, sizeof (local_builder));
+}
+
+/**
+ * epc_app_filter_builder_clear:
+ * @builder: an #EpcAppFilterBuilder
+ *
+ * Clear @builder, freeing any internal state in it. This will not free the
+ * top-level storage for @builder itself, which is assumed to be allocated on
+ * the stack.
+ *
+ * If called on an already-cleared #EpcAppFilterBuilder, this function is
+ * idempotent.
+ *
+ * Since: 0.1.0
+ */
+void
+epc_app_filter_builder_clear (EpcAppFilterBuilder *builder)
+{
+  EpcAppFilterBuilderReal *_builder = (EpcAppFilterBuilderReal *) builder;
+
+  g_return_if_fail (_builder != NULL);
+
+  g_clear_pointer (&_builder->paths_blacklist, g_ptr_array_unref);
+  g_clear_pointer (&_builder->oars, g_hash_table_unref);
+}
+
+/**
+ * epc_app_filter_builder_new:
+ *
+ * Construct a new #EpcAppFilterBuilder on the heap. This is intended for
+ * language bindings. The returned builder must eventually be freed with
+ * epc_app_filter_builder_free(), but can be cleared zero or more times with
+ * epc_app_filter_builder_clear() first.
+ *
+ * Returns: (transfer full): a new heap-allocated #EpcAppFilterBuilder
+ * Since: 0.1.0
+ */
+EpcAppFilterBuilder *
+epc_app_filter_builder_new (void)
+{
+  g_autoptr(EpcAppFilterBuilder) builder = NULL;
+
+  builder = g_new0 (EpcAppFilterBuilder, 1);
+  epc_app_filter_builder_init (builder);
+
+  return g_steal_pointer (&builder);
+}
+
+/**
+ * epc_app_filter_builder_copy:
+ * @builder: an #EpcAppFilterBuilder
+ *
+ * Copy the given @builder to a newly-allocated #EpcAppFilterBuilder on the
+ * heap.
+ *
+ * Returns: (transfer full): a copy of @builder
+ * Since: 0.1.0
+ */
+EpcAppFilterBuilder *
+epc_app_filter_builder_copy (EpcAppFilterBuilder *builder)
+{
+  EpcAppFilterBuilderReal *_builder = (EpcAppFilterBuilderReal *) builder;
+  g_autoptr(EpcAppFilterBuilder) copy = NULL;
+  EpcAppFilterBuilderReal *_copy;
+
+  g_return_val_if_fail (builder != NULL, NULL);
+
+  copy = epc_app_filter_builder_new ();
+  _copy = (EpcAppFilterBuilderReal *) copy;
+
+  _copy->paths_blacklist = g_ptr_array_ref (_builder->paths_blacklist);
+  _copy->oars = g_hash_table_ref (_builder->oars);
```

Comment:
eg. this assumes `oars` is non-NULL

Reply:
Yeah, this `copy` function was a quick addition which I probably should have put more thought into. I only added it to enable the type to be boxed. It should probably safely copy `NULL` values.

Reply:
Modified to safely copy cleared (`NULL` elements) builders.